docs/source/best_practices.rst

```diff
@@ -0,0 +1,187 @@
+
+    def get_identifier_name(node: cst.CSTNode) -> Optional[str]:
+        if m.matches(node, m.Name()):
+            assert isinstance(node, cst.Name)
+            return node.value
+        return None
+
+Consider writing instead::
+
+    def get_identifier_name(node: cst.CSTNode) -> Optional[str]:
+        return (
+            cst.ensure_type(node, cst.Name).value
+            if m.matches(node, m.Name())
+            else None
+        )
+
+If you find yourself attempting to manually traverse a tree using ``isinstance``,
+you can often rewrite your code using visitor methods instead. Nested instance checks
+can instead be unrolled into simpler code using visitors with matcher decorators. For
+example, instead of writing the following::
+
+    class CountBazFoobarArgs(cst.CSTVisitor):
+        """
+        Given a set of function names, count how many arguments to those function
+        calls are the identifiers "baz" or "foobar".
+        """
+
+        def __init__(self, functions: Set[str]) -> None:
+            super().__init__()
+            self.functions: Set[str] = functions
+            self.arg_count: int = 0
+
+        def visit_Call(self, node: cst.Call) -> None:
+            # See if the call itself is one of our functions we care about
+            if isinstance(node.func, cst.Name) and node.func.value in self.functions:
+                # Loop through each argument
+                for arg in node.args:
+                    # See if the argument is an identifier matching what we want to count
+                    if isinstance(arg.value, cst.Name) and arg.value.value in {"baz", "foobar"}:
+                        self.arg_count += 1
+                        
+Consider writing instead::
+
+    class CountBazFoobarArgs(m.MatcherDecoratableVisitor):
+        """
+        Given a set of function names, count how many arguments to those function
+        calls are the identifiers "baz" or "foobar".
+        """
+
+        def __init__(self, functions: Set[str]) -> None:
+            super().__init__()
+            self.functions: Set[str] = functions
+            self.arg_count: int = 0
+            self.call_stack: List[str] = []
+
+        def visit_Call(self, node: cst.Call) -> None:
+            # Store all calls in a stack
+            if m.matches(node.func, m.Name()):
+                self.call_stack.append(cst.ensure_type(node.func, m.Name).value)
+
+        def leave_Call(self, original_node: cst.Call) -> None:
+            # Pop the latest call off the stack
+            if m.matches(node.func, m.Name()):
+                self.call_stack.pop()
+
+        @m.visit(m.Arg(m.Name("baz") | m.Name("foobar")))
+        def _count_args(self, node: cst.Arg) -> None:
+            # See if the most shallow call is one we're interested in, so we can
+            # count the args we care about only in calls we care about.
+            if self.call_stack[-1] in self.functions:
+                self.arg_count += 1
```

Comment:
This example collected call stack and that made it complex.
Looks like we can do it in a much simpler way with a matcher check?
```
def visit_Call(self, node: cst.Call) -> None:
    if m.matches(node, m.Call(func=Name(), value=m.OneOf(*[m.Name(s) for s in self. functions]))):
        self.arg_count += [m.matches(arg.value, m.OneOf(m.Name("baz"), m.Name("foobar"))) for arg in node.args]
```

Reply:
The point of this example is showing how to unroll a set of instance checks into more than one visitor method. Your proposed solution does get that point across.

Reply:
I understand the point but the new code doesn't seem simpler as it claims to be.

In old code, all logic were inside visit_Call, which is easy to follow. In new code, the reader needs to figure out the orders and consequences of those functions (visit_Call, leave_Call and _count_arg), which seems harder to read.
I hope the example can present the `simpler` better.

Reply:
I completely disagree. Your proposed example uses a lot of Python features to make the least amount of code, but the density of required reading is high. I have trouble following it and I only can tell what it does because it proposes replacing other code with the same effect. It is also broken (you can't add a list to an int). It also continues to do manual traversal of the tree, which the recommended practice says to avoid. It also doesn't help demonstrate how somebody might spread logic across multiple visitors using state, which is precisely one motivation for moving to visitors instead of matchers/deep_replace. Basically, it is not only not a good example for the practice discussed on the page, but I would not view it is a good practice in general.

Reply:
I'm not saying we should use my example. Just to see if there is a better example can demonstrate it's a lot simpler to use visitor functions.

Reply:
To make it clear, my second comment wasn't proposing to use the example in my first comment. I tried to propose simplifying your example by comparing your before and after codes to demonstrate the after codes wasn't simpler. Sorry about the confusion if it wasn't clear enough.
For `visit_<CSTNode>()` function with consequences and orders tracked by class attributes (e.g. the stack variable or just a boolean in some other use cases), it's usually less readable because user need to figure out the orders and the consequences. I think the matcher decorator pattern like `call_if_inside` + `visit_<CSTNode>` is a better way to handle it. It keep the order of logic and consequences in the same place which is a lot more readable.

```
@call_if_inside(m.Call(func=m.OneOf(*[m.Name(s) for s in self.functions])))
def visit_Arg(node: cst.Arg):
    if m.matches(node.value, m.OneOf(m.Name("baz"), m.Name("foobar"))):
        self.count += 1
```